### PR TITLE
images/builder: Allow GCB builds from arbitrary build directories

### PR DESCRIPTION
**Pull request by justaugustus**
- Remove `cdToRootDir`
    This function was not actually cd-ing into directories
- Add `--build-dir` flag to support builds that require the entire
   repo/workspace to be uploaded to GCB
- Add a validation method to ensure config directory and `cloudbuild.yaml`
   exist before submitting build

Signed-off-by: Stephen Augustus <saugustus@vmware.com>

/assign @Katharine @dims @tpepper 
cc: @alexeldeib @kubernetes/release-engineering 
/sig release
/area release-eng
/milestone v1.17
/priority important-soon

---

After wiring k/release up to do image building/pushing via post-submit (https://github.com/kubernetes/test-infra/pull/14711, https://github.com/kubernetes/test-infra/pull/14735, https://github.com/kubernetes/release/pull/895, https://github.com/kubernetes/release/pull/896), I noticed that we had a few build failures:

`https://prow.k8s.io/view/gcs/kubernetes-jenkins/logs/post-release-push-image-k8s-cloud-builder/1182793236758925320`:
```
Created [https://cloudbuild.googleapis.com/v1/projects/k8s-staging-release-test/builds/75d28263-9671-4359-854e-2535a29ad3b0].
Logs are available at [https://console.cloud.google.com/gcr/builds/75d28263-9671-4359-854e-2535a29ad3b0?project=634027639865].
----------------------------- REMOTE BUILD OUTPUT ------------------------------
ERROR: (gcloud.builds.submit) build 75d28263-9671-4359-854e-2535a29ad3b0 completed with status "FAILURE"
starting build "75d28263-9671-4359-854e-2535a29ad3b0"

FETCHSOURCE
Fetching storage object: gs://k8s-staging-release-test-gcb/source/1570834875.37-297cf4b31fb547218dc02ad9004339f8.tgz#1570834875611654
Copying gs://k8s-staging-release-test-gcb/source/1570834875.37-297cf4b31fb547218dc02ad9004339f8.tgz#1570834875611654...
/ [0 files][    0.0 B/166.4 KiB]                                                
/ [1 files][166.4 KiB/166.4 KiB]                                                
Operation completed over 1 objects/166.4 KiB.                                    
BUILD
Already have image (with digest): gcr.io/cloud-builders/docker
unable to prepare context: unable to evaluate symlinks in Dockerfile path: lstat /workspace/Dockerfile: no such file or directory
ERROR
ERROR: build step 0 "gcr.io/cloud-builders/docker" failed: exit status 1
--------------------------------------------------------------------------------
```

`https://prow.k8s.io/view/gcs/kubernetes-jenkins/logs/post-release-push-image-deb-builder/1182793236758925319`: 
```
Created [https://cloudbuild.googleapis.com/v1/projects/k8s-staging-release-test/builds/42fb12ed-fff8-4e50-8cc2-21037cbf38e1].
Logs are available at [https://console.cloud.google.com/gcr/builds/42fb12ed-fff8-4e50-8cc2-21037cbf38e1?project=634027639865].
----------------------------- REMOTE BUILD OUTPUT ------------------------------
starting build "42fb12ed-fff8-4e50-8cc2-21037cbf38e1"

FETCHSOURCE
Fetching storage object: gs://k8s-staging-release-test-gcb/source/1570834875.55-c6f9168b8c5c4e239177f86f7161df02.tgz#1570834875702092
Copying gs://k8s-staging-release-test-gcb/source/1570834875.55-c6f9168b8c5c4e239177f86f7161df02.tgz#1570834875702092...
/ [0 files][    0.0 B/166.4 KiB]                                                
/ [1 files][166.4 KiB/166.4 KiB]                                                
Operation completed over 1 objects/166.4 KiB.                                    
BUILD
Starting Step #0
Step #0: Pulling image: golang:1.13
Step #0: 1.13: Pulling from library/golang
Step #0: Digest: sha256:68f8870ee1723cafd45bed29414fbaa151e9bf2bba369c8b4436c08a18907012
Step #0: Status: Downloaded newer image for golang:1.13
Step #0: docker.io/library/golang:1.13
Step #0: build .: cannot find module for path .
Finished Step #0
ERROR
ERROR: build step 0 "golang:1.13" failed: exit status 1
--------------------------------------------------------------------------------

ERROR: (gcloud.builds.submit) build 42fb12ed-fff8-4e50-8cc2-21037cbf38e1 completed with status "FAILURE"
```

`https://prow.k8s.io/view/gcs/kubernetes-jenkins/logs/post-release-push-image-rpm-builder/1182793236758925321`: 
```
Created [https://cloudbuild.googleapis.com/v1/projects/k8s-staging-release-test/builds/133ed19d-2e9d-41ae-92f1-0a04766a25e3].
Logs are available at [https://console.cloud.google.com/gcr/builds/133ed19d-2e9d-41ae-92f1-0a04766a25e3?project=634027639865].
----------------------------- REMOTE BUILD OUTPUT ------------------------------
ERROR: (gcloud.builds.submit) build 133ed19d-2e9d-41ae-92f1-0a04766a25e3 completed with status "FAILURE"
starting build "133ed19d-2e9d-41ae-92f1-0a04766a25e3"

FETCHSOURCE
Fetching storage object: gs://k8s-staging-release-test-gcb/source/1570834874.08-db53a5ba6bd64a5d8bbfe614d8634184.tgz#1570834874408813
Copying gs://k8s-staging-release-test-gcb/source/1570834874.08-db53a5ba6bd64a5d8bbfe614d8634184.tgz#1570834874408813...
/ [0 files][    0.0 B/166.4 KiB]                                                
/ [1 files][166.4 KiB/166.4 KiB]                                                
Operation completed over 1 objects/166.4 KiB.                                    
BUILD
Already have image (with digest): gcr.io/cloud-builders/docker
unable to prepare context: unable to evaluate symlinks in Dockerfile path: lstat /workspace/Dockerfile: no such file or directory
ERROR
ERROR: build step 0 "gcr.io/cloud-builders/docker" failed: exit status 1
--------------------------------------------------------------------------------
```

Basically what's happening is the builder is looking for assets in `/workspace` and can't find them.
This was happening because the `os.Chdir` command was in a separate function scope, so it was ultimately not being respected. This was further obfuscated by the fact that the builder in its' current state specifies the `cloudbuild.yaml` while you're outside of the image directory.

I tried testing this locally with commands close to what would be running in CI and all seems to be working...

Build the image builder from the root of k/test-infra:
```shell
CGO_ENABLED=0 \
GOOS=linux \
GOARCH=amd64 \
GO111MODULE=on \
GOPROXY=https://proxy.golang.org \
GOSUMDB=sum.golang.org \
go build -o=./images/builder/builder ./images/builder
```

Submitting an image push from the root of k/release:
```shell
time /home/augustus/go/src/k8s.io/test-infra/images/builder/builder --project=kubernetes-release-test --scratch-bucket=gs://kubernetes-release-test-gcb --env-passthrough=PULL_BASE_REF build/debs/
2019/10/12 01:36:51 working directory: /home/augustus/go/src/k8s.io/release/build/debs
2019/10/12 01:36:51 Creating source tarball at /tmp/221087232...
2019/10/12 01:36:54 Uploading /tmp/221087232 to gs://kubernetes-release-test-gcb/source/9464065b-2e94-4ebd-84e7-591dff4d38ef.tgz...
Copying file:///tmp/221087232 [Content-Type=application/octet-stream]...
- [1 files][ 57.4 MiB/ 57.4 MiB]    1.2 MiB/s                                   
Operation completed over 1 objects/57.4 MiB.                                     
2019/10/12 01:37:37 Running build jobs...
2019/10/12 01:37:37 No variants.yaml, starting single build job...
Created [https://cloudbuild.googleapis.com/v1/projects/kubernetes-release-test/builds/47e371f8-d6d9-4370-8a56-567236fe103a].
Logs are available at [https://console.cloud.google.com/gcr/builds/47e371f8-d6d9-4370-8a56-567236fe103a?project=648026197307].
------------------------------------------------------------------------------------------------------ REMOTE BUILD OUTPUT -------------------------------------------------------------------------------------------------------
starting build "47e371f8-d6d9-4370-8a56-567236fe103a"

FETCHSOURCE
Fetching storage object: gs://kubernetes-release-test-gcb/source/1570858658.2-904011a4abf44c4eb7cfc1663efaa590.tgz#1570858658617748
Copying gs://kubernetes-release-test-gcb/source/1570858658.2-904011a4abf44c4eb7cfc1663efaa590.tgz#1570858658617748...
\ [1 files][ 57.4 MiB/ 57.4 MiB]                                                
Operation completed over 1 objects/57.4 MiB.                                     
BUILD

<snip>

DONE
----------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------

ID                                    CREATE_TIME                DURATION  SOURCE                                                                                     IMAGES                                                                              STATUS
47e371f8-d6d9-4370-8a56-567236fe103a  2019-10-12T05:37:38+00:00  2M16S     gs://kubernetes-release-test-gcb/source/1570858658.2-904011a4abf44c4eb7cfc1663efaa590.tgz  gcr.io/kubernetes-release-test/deb-builder:v20191012-v0.1.3-124-ga49231d (+2 more)  SUCCESS
2019/10/12 01:39:56 Finished.

real	3m5.147s
user	0m6.419s
sys	0m1.151s
```

### CODE REVIEW

**Comment by Katharine** [comment]
/hold

This will break a number of existing jobs — I think all the test-infra uploads, which both rely on the `cdToRootDir` behaviour as well as in some cases the fact that we do actually upload the entire repo and not just the path that was given (because of go.mod files, mostly).

I'm not sure it's better to change this behaviour, given doing so potentially creates the need for submodules or substantial refactoring in monorepos (as you already experienced, as a result of my faulty recollection of how this code actually works — sorry about that).

**Comment by justaugustus** [comment]
@Katharine -- thanks for the chat earlier today to decide on approach!

I've tweaked the builder so it now supports:
- builds where all assets are in the image directory
- builds where all assets are in the root of the workspace, including `cloudbuild.yaml`
- builds where some assets (like `go.mod`, `go.sum`) are in the root of the workspace, but `cloudbuild.yaml` is in a subdirectory (a la the jobs in `test-infra-trusted.yaml`)

I've tested this on a few of the `k/release` jobs as well as the builder image itself, and I've also updated each of the jobs in `test-infra-trusted.yaml` to use `--build-at-root`, so this should be good to merge now. :)

File: images/builder/main.go [hunk 1]
```diff
@@ -257,27 +269,69 @@ func mergeMaps(maps ...map[string]string) map[string]string {
...
+				log.Fatalf("Failed to git rev-parse workspace: %v\n", err)
+
+				chdirErr := os.Chdir(strings.TrimSpace(string(rootDir)))
+				if chdirErr != nil {
+					log.Fatalf("Failed to cd to root of workspace: %v\n", chdirErr)
+				}
+			}
+		}
+	} else {
+		o.cloudbuildPath = "cloudbuild.yaml"
```
Comment (Katharine) [inline images/builder/main.go]: I'm not sure that this `cloudbuildPath` option is what we want. There are essentially two concepts we have:

- Where are our config files located?
- What are we sending to GCB?

Notably, there are _two_ config files - `cloudbuild.yaml` and `variants.yaml` - which should be together, but not necessarily in the root. I think that this should be the path to that _directory_, and the paths to the relevant files can be assembled by appending to it later.

The other concept - what we send to GCB - either matches the config dir (by default), or is "the root", which we are really interpreting as "the current working directory" (which is a slightly awkward concept).

I still wonder whether it would make sense to instead of having `--build-at-root`, instead have `--upload-dir`, which would be equivalent to the current build-at-root with `--upload-dir=.` Then you would invoke the builder as `image-builder --upload-dir=. ./images/builder/` or something. I think in this case, both `upload-dir` and the position argument should be relative to the _current_ working directory, rather than eachother (since there is not actually any requirement that the config be part of the build).

**Comment by justaugustus** [comment]
@Katharine -- I updated the builder based on your suggestions and tested it on a few k/release jobs, including one with a `variants.yaml` (in https://github.com/kubernetes/release/pull/900), as well as the builder itself.

I've also re-updated the test-infra-trusted jobs to use the `--build-dir` flag.

File: images/builder/main.go [hunk 2]
```diff
@@ -257,27 +269,69 @@ func mergeMaps(maps ...map[string]string) map[string]string {
...
 
+	if o.buildAtRoot {
+		o.cloudbuildPath = path.Join(o.imageDirectory, "cloudbuild.yaml")
+
+		if bazelWorkspace := os.Getenv("BUILD_WORKSPACE_DIRECTORY"); bazelWorkspace != "" {
+			if err := os.Chdir(bazelWorkspace); err != nil {
+				log.Fatalf("Failed to chdir to bazel workspace (%s): %v", bazelWorkspace, err)
+			}
+
+			cmd := exec.Command("git", "rev-parse", "--show-toplevel")
```
[the remark this answers is not in the document]
Comment (justaugustus) [reply images/builder/main.go]: Done.

File: images/builder/main.go [hunk 3]
```diff
@@ -257,27 +269,69 @@ func mergeMaps(maps ...map[string]string) map[string]string {
...
+
+	imageDirErr := o.validateImageDir()
+	if imageDirErr != nil {
+		log.Fatalf("Could not validate image directory: %v", imageDirErr)
 	}
 
+	if o.buildAtRoot {
+		o.cloudbuildPath = path.Join(o.imageDirectory, "cloudbuild.yaml")
+
+		if bazelWorkspace := os.Getenv("BUILD_WORKSPACE_DIRECTORY"); bazelWorkspace != "" {
```
[the remark this answers is not in the document]
Comment (justaugustus) [reply images/builder/main.go]: Done.

File: images/builder/main.go [hunk 4]
```diff
@@ -153,7 +170,7 @@ func runSingleJob(o options, jobName, uploaded, version string, subs map[string]
 type variants map[string]map[string]string
 
 func getVariants(o options) (variants, error) {
-	content, err := ioutil.ReadFile(path.Join(o.imageDirectory, "variants.yaml"))
+	content, err := ioutil.ReadFile("variants.yaml")
```
Comment (Katharine) [inline images/builder/main.go]: Is this right? We should be looking for this in the config dir, but at this point we will be in the build dir.

File: images/builder/main.go [hunk 5]
```diff
@@ -257,30 +280,69 @@ func mergeMaps(maps ...map[string]string) map[string]string {
...
 	flag.Parse()
+
 	if flag.NArg() < 1 {
-		_, _ = fmt.Fprintln(os.Stderr, "expected an image directory to be provided")
+		_, _ = fmt.Fprintln(os.Stderr, "expected a config directory to be provided")
 		os.Exit(1)
 	}
-	o.imageDirectory = flag.Arg(0)
+
+	o.configDir = strings.TrimSuffix(flag.Arg(0), "/")
```
Comment (Katharine) [inline images/builder/main.go]: Why is this change necessary?

Comment (justaugustus) [reply images/builder/main.go]: @Katharine -- I don't think it's strictly necessary, but it's defensive against any weirdness with path joins, double `/`s, etc.

Comment (dougm) [reply images/builder/main.go]: `filepath.Clean` takes care of that and more: https://golang.org/pkg/path/filepath/#Clean

**Comment by Katharine** [comment]
This PR appears to have rendered the image builder unable to build itself, reverting
